whisper/wnode/main.go

```diff
@@ -0,0 +1,495 @@
+// Copyright 2016 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// This is a simple Whisper node.
+// It also implements a command line chat between the peers sharing the same credentials.
+
+package main
+
+import (
+	"bufio"
+	"crypto/ecdsa"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/cmd/utils"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/console"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/logger"
+	"github.com/ethereum/go-ethereum/logger/glog"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/nat"
+	whisper "github.com/ethereum/go-ethereum/whisper/whisperv5"
+	"golang.org/x/crypto/pbkdf2"
+)
+
+var done chan struct{}
+var server *p2p.Server
+var shh *whisper.Whisper
+var asymKey, nodeid *ecdsa.PrivateKey // used for asymmetric decryption and signing
+var pub *ecdsa.PublicKey              // used for asymmetric encryption
+var symKey []byte                     // used for symmetric encryption
+var filterID uint32
+var topic whisper.TopicType
+var pow float64 = whisper.MinimumPoW
+var ttl uint32 = 30
+var workTime uint32 = 5
+var ipAddress, enode, salt, topicStr, pubStr, NodeIdFile string
+var input *bufio.Reader = bufio.NewReader(os.Stdin)
+
+var bootstrapNode bool // does not actively connect to peers, wait for incoming connections
+var daemonMode bool    // only forward messages, neither send nor track
+var testMode bool      // predefined password, topic, ip and port
+var echoMode bool      // shows params, etc.
+var isAsymmetric bool
+
+var (
+	argNameHelp       = "-h"
+	argNameBootstrap  = "-b"
+	argNameDaemon     = "-d"
+	argNameAsymmetric = "-a"
+	argNameTest       = "-test"
+	argNameEcho       = "-echo"
+	argNameWorkTime   = "-work"
+	argNameTTL        = "-ttl"
+	argNameIP         = "-ip"
+	argNameTopic      = "-topic"
+	argNamePoW        = "-pow"
+	argNameSalt       = "-salt"
+	argNamePub        = "-pub"
+	argNameEnode      = "-enode"
+	argNameIdSrc      = "-idfile"
+	quitCommand       = "~Q"
+)
```

Comment:
Please use [package flag](https://golang.org/pkg/flag/) to handle command line flags. It will remove a lot of code from this command.